File: analyzers/tests/SonarAnalyzer.Test/TestCases/UnusedStringBuilder.vb
Imports System.Text

Public Class Program
    Public Function NotUsed(ByVal builder As StringBuilder) As StringBuilder
        Dim builder1 As StringBuilder = GetStringBuilder()  ' Compliant
        Dim builder2 = New StringBuilder()                  ' Compliant
        ExternalMethod(builder2)
        Dim builder3 As StringBuilder = New StringBuilder() ' Compliant
        builder3.ToString()
        Dim builder4 = New StringBuilder()                  ' Compliant
        Dim builder5 As StringBuilder = New StringBuilder() ' Noncompliant
        Dim builder6 = New StringBuilder()                  ' Noncompliant
        Dim builder7 As StringBuilder = New StringBuilder() ' Noncompliant
        builder7.Append(builder4.ToString())
        Dim builder8 As StringBuilder = New StringBuilder() ' Compliant
        builder8.Append("&").ToString()

        Dim builderInLine1 As StringBuilder = New StringBuilder(), builderInLine2 As StringBuilder = New StringBuilder()
        '   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                           Noncompliant
        '                                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^    Noncompliant@-1

        Dim sb1, sb2 ' Compliant
        Dim sb3, sb4 As StringBuilder = New StringBuilder() ' Error [BC30671]: Explicit initialization is not permitted with multiple variables declared with a single type specifier.

        Dim sb5 As StringBuilder = New StringBuilder() ' Compliant
        Dim sb6 As StringBuilder = New StringBuilder() ' Noncompliant
        sb6 = sb5

        Dim builderCfg As StringBuilder = New StringBuilder() ' FN (requires use of cfg with significant impact on performance)
        If False Then
            builderCfg.ToString()
        End If

        Dim builderCalls As StringBuilder = New StringBuilder() ' Noncompliant
        builderCalls.Append("Append")
        builderCalls.AppendLine("AppendLine")
        builderCalls.Replace("a", "b")
        builderCalls.Clear()

        Dim builderCalls2 As StringBuilder = New StringBuilder() ' Compliant
        builderCalls2.Remove(builderCalls2.Length - 1, 1)

        Dim mySbField As StringBuilder = New StringBuilder()    ' Noncompliant
        Dim [myClass] As [MyClass] = New [MyClass]()
        [myClass].mySbField.ToString()

        Dim builderReturn = New StringBuilder()                 ' Compliant
        Return builderReturn
    End Function

    Public Function GetStringBuilder() As StringBuilder     ' Compliant
        Return New StringBuilder()
    End Function

    Public Sub ExternalMethod(ByVal builder As StringBuilder) ' Compliant
    End Sub

    Public Function AnotherMethod() As String
        Dim builder = New StringBuilder() ' Compliant
        Return $"{builder} is ToStringed here"
    End Function

    Public ReadOnly Property MyProperty As String
        Get
            Dim builder1 = New StringBuilder()              ' Noncompliant
            Dim builder2 = New StringBuilder()              ' Compliant
            Return builder2.ToString()
        End Get
    End Property

    Private myField As StringBuilder = New StringBuilder() ' Compliant
End Class

Public Class [MyClass]
    Public mySbField As StringBuilder = New StringBuilder()
End Class

' https://github.com/SonarSource/sonar-dotnet/issues/8809
Public Class GH8809
    Public Function ToString() As String
        Dim builder As New StringBuilder() ' FN
        builder.AppendLine("value")

        Return String.Empty
    End Function
End Class

File: analyzers/tests/SonarAnalyzer.Test/TestCases/UnusedStringBuilder.cs
using System.Text;

public class Program
{
    public StringBuilder MyMethod(StringBuilder builder, string myString) // Compliant
    {
        StringBuilder sb1 = GetStringBuilder(); // Compliant

        var sb2 = new StringBuilder(); // Compliant
        ExternalMethod(sb2);

        var sb3 = new StringBuilder(); // Compliant
        LocalMethod(sb3);

        var sb4 = new StringBuilder(); // Compliant

        StringBuilder sb5 = new StringBuilder(); // Noncompliant {{Remove this "StringBuilder"; ".ToString()" is never called.}}
//                    ^^^^^^^^^^^^^^^^^^^^^^^^^
        var sb6 = new StringBuilder(); // Noncompliant
//          ^^^^^^^^^^^^^^^^^^^^^^^^^

        StringBuilder sb7 = new StringBuilder(); // Noncompliant
        sb7.Append(sb4.ToString());

        StringBuilder sb8 = new StringBuilder(); // Compliant
        StringBuilder sb9 = new StringBuilder(); // Noncompliant
        sb9 = sb8;

        StringBuilder sb10 = new StringBuilder(); // Noncompliant FP see https://github.com/SonarSource/sonar-dotnet/issues/6747
        sb10.ToStringAndFree();

        if (true)
        {
            if (true)
            {
                if (true)
                {
                    if (true)
                    {
                        if (true)
                        {
                            var buildernested = new StringBuilder(); // Noncompliant
                        }
                    }
                }
            }
        }

        (StringBuilder, StringBuilder) builderTuple = (new StringBuilder(), new StringBuilder()); // FN

        StringBuilder builderInLine1 = new StringBuilder(), builderInLine2 = new StringBuilder();
//                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                          Noncompliant
//                                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^    Noncompliant@-1


        StringBuilder builderCfg = new StringBuilder(); // FN (requires use of cfg with significant impact on performance)
        if (false)
        {
            builderCfg.ToString();
        }

        StringBuilder builderCalls = new StringBuilder(); // Noncompliant
        builderCalls.Append("Append");
        builderCalls.AppendLine("AppendLine");
        builderCalls.Replace("\r\n", "\n");
        builderCalls.Clear();
        builderCalls.GetType();

        StringBuilder mySbField = new StringBuilder(); // Noncompliant
        MyClass myClass = new MyClass();
        myClass.mySbField.ToString();

        var builderReturn = new StringBuilder(); // Compliant
        return builderReturn;

        void LocalMethod(StringBuilder local)
        {
            local.ToString();
        }
    }

    public StringBuilder GetStringBuilder() => // Compliant
        new StringBuilder();

    public void ExternalMethod(StringBuilder builder) { } // Compliant

    public string AnotherMethod()
    {
        var builder = new StringBuilder(); // Compliant
        return $"{builder} is ToStringed here";
    }

    public void Scoping()
    {
        {
            var sb = new StringBuilder(); // Noncompliant
        }
        {
            var sb = new StringBuilder();
            sb.ToString();
        }
    }

    public string CompoundAdd()
    {
        var sb = new StringBuilder();   // Compliant
        sb.Append("Lorem ipsum");
        string ret = "";
        ret += sb;
        return ret;
    }

    public string MyProperty
    {
        get
        {
            var builder1 = new StringBuilder(); // Noncompliant
            var builder2 = new StringBuilder(); // Compliant
            return builder2.ToString();
        }
    }

    private StringBuilder myField = new StringBuilder(); // Compliant
}

public class MyClass
{
    public StringBuilder mySbField = new StringBuilder();
}

public static class StringBuilderExtensions
{
    public static string ToStringAndFree(this StringBuilder sb) => string.Empty;
}

// https://github.com/SonarSource/sonar-dotnet/issues/7324
public class Repro_7324
{
    public string Concat_Prefix()
    {
        var sb = new StringBuilder();   // Compliant
        sb.Append("Lorem ipsum");
        var ret = "Prefix: " + sb;
        return ret;
    }

    public string Concat_Infix()
    {
        var sb = new StringBuilder();   // Compliant
        sb.Append("Lorem ipsum");
        var ret = "Prefix: " + sb + " suffix";
        return ret;
    }

    public string Concat_Suffix()
    {
        var sb = new StringBuilder();   // Compliant
        sb.Append("Lorem ipsum");
        var ret = sb + " suffix";
        return ret;
    }

    public string Concat_OutsideDeclaration()
    {
        var sb = new StringBuilder();   // Compliant
        sb.Append("Lorem ipsum");
        string ret;
        ret = "Prefix: " + sb;
        return ret;
    }
}
